src/fsharp/pars.fsy

```diff
@@ -3499,7 +3520,7 @@ declExpr:
       { let mMatch = rhs parseState 1
         let mWith, (clauses, mLast) = $3
         let spBind = DebugPointAtBinding.Yes(unionRanges mMatch mWith)
-        SynExpr.Match (spBind, $2, clauses, unionRanges mMatch mLast) }
+        SynExpr.Match (mMatch, spBind, $2, mWith, clauses, unionRanges mMatch mLast) }
```

Comment:
@nojar What problem does a separate `match` field solve here? The keyword starts at the same range as the whole `match` expression and is single token of a known width. Given each added range is rather a big struct, it seems a bit wasteful to add ranges that can be calculated from the nodes.

Maybe we could add some helpers that would calculate ranges for keywords from their positions instead of storing them in tree?

Reply:
Feels like we are talking in circles here:
![image](https://user-images.githubusercontent.com/2621499/142587135-6d14aab9-b6af-45df-9a92-5d079efe5d45.png)

I already expressed why ranges are preferred over calculations with positions.
@dsyme what do you think?

Reply:
This time I'm suggesting this calculation to be the details of implementation in FCS, while keeping range properties that Fantomas could work with. 🙂

Reply:
Well, I'm slowly getting on board with this.
Although, is there a way to get a position inside the parser?
Places where I now use `rhs parseState 1`, is there something similar that just gets the start position of 1?

Reply:
@auduchinok would that then look something like:
```fsharp
#r "nuget: FSharp.Compiler.Service"

open FSharp.Compiler.Text
open FSharp.Compiler.Text.Range
open FSharp.Compiler.Text.Position

type NodeTrivia =
    { StartPos: pos
      FileName: string }

    member this.Range =
        mkRange this.FileName this.StartPos (mkPos this.StartPos.Line (this.StartPos.Column + 4))

let myTrivia =
    { StartPos = mkPos 1 0
      FileName = "MyFile" }

myTrivia.Range
```
?

Reply:
@nojaf Yes the idea for getting is roughly like this. I'm also trying to find a better way to keep the keyword position only and to get the file name from the overall node range. Something like this could work:
```fsharp
let getKeywordRange fileName length startPos =
    mkRange fileName startPos (mkPos startPos.Line (startPos.Column + length))

type SynExprMatchTrivia =
    { WithPos: pos }

    member this.GetMatchKeywordRange(exprRange: range) =
        getKeywordRange exprRange.FileName exprRange.Start FSharpKeywords.Match.Length

    member this.GetWithKeywordRange(exprRange: range) =
        getKeywordRange exprRange.FileName this.WithPos FSharpKeywords.With.Length
```

Reply:
I would prefer a range than a position.  The position on its own is of no use, we would always reconstruct the range.  These particular nodes are of no intrinsic significance when it comes to memory performance